### DISCUSSION TITLE
common lisp?

### DISCUSSION THREAD

**Discussion by masinter** [Interlisp/medley#5] [Q&A]
Bob said he last worked on common lisp and had a marked-up copy of CLTL2 at his office with all the problems.

So if we can, fixing 'up common lisp support is a good long-term goal, but just knowing what we have will be helpful.

**Comment by Anzus** [comment]
If Bob doesn't have it anymore I have an on-disk copy of a tape he sent me of his work to make Medley CLTL2 compliant. A big issue was the CL package nickname changed from LISP to COMMON-LISP.

**Comment by Anzus** [comment]
Oh, also this predates ANSI. So one question is whether we release CLTL1, CLTL2, or ANSI...

**Comment by masinter** [comment]
We release what we have. I’m just thinking about release notes

**Comment by Kirtai** [comment]
Re #47, I found more information int the Medley release notes (not in any manual I could find) and have added it here so it doesn't get lost.

CL is CLtL1.
The CL interactor uses the LISP readtable and the USER package.
The XCL interactor uses the XCL readtable and the XCL-USER package.
The LISP readtable is pure CLtL1
For XCL: 

> "This readtable is the same as LISP, except that the characters with ASCII codes 1 thru 26 have white-space (SEPRCHAR) syntax . This readtable is intended for use in File Manager files, so that font information can be encoded on the file"

They both accept commands like "undo" which doesn't quite match with the above.

**Comment by masinter** [comment]
What we have is called "Medley 3.5". we're not sure what changes were made while Venue was active (prior to 2009).
The Envos files we have are from 1993 and contains releases for Medley 1.0 and Medly 2.0 (and maybe a few in between).
Medley 3.5 wasn't officially 'released', though, so that's why we don't know what we have exactly and the "restore" part of the project is important. 1993-2009 seems like a long time, and a long time ago.

Can we build a Medley 3.5 Release with proper release notes?

**Comment by masinter** [comment]
> If Bob doesn't have it anymore I have an on-disk copy of a tape he sent me of his work to make Medley CLTL2 compliant. A big issue was the CL package nickname changed from LISP to COMMON-LISP.

That would be quite useful; if we can link up the dates on his notes to the dates of the common lisp implementation files.

**Comment by pamoroso** [comment]
I'm putting together this list of missing features and differences of Medley Common Lisp with ANSI X3J13.

The entries are grouped by category in sections with the same headings of the corresponding HyperSpec chapters. Unless otherwise noted the listed features are missing from Medley.

The list, which I'll expand and revise, is based on my own experimentation and other sources such as discussions #5 and #1473 and issue #609. Should it be span off into its own issue for easier reference and maintenance?


## 2. Syntax

* `#|` `|#`


## 3. Evaluation and Compilation

- `declaim`
- `load-time-value`
- `symbol-macrolet`
- `values` is not supported as a location specifier for `setf`


## 6. Iteration

- `loop` doesn't support iterating over packages


## 7. Objects

Docstrings of generic functions are ignored.


## 9. Conditions

- `handler-case` is in package `CONDITIONS` (used by `XCL`) instead of `CL`
- `use-value` doesn't allow to enter a value interactively


## 11. Packages

I'm not sure what issue @Anzus [pointed out](https://github.com/orgs/Interlisp/discussions/5#discussioncomment-198669) about the nicknames of package `CL`. In Medley the `COMMON-LISP` package seems to have the nickname `CL` as the HyperSpec requires in [section 11.1.2.1 The COMMON-LISP Package](https://www.lispworks.com/documentation/HyperSpec/Body/11_aba.htm).

- the `COMMON-LISP` and `COMMON-LISP-USER` packages are missing but their Medley equivalents are `LISP` (nick: `CL`) and `USER`
- `defpackage` declarations like `(defpackage :pkgname (:use ...) (:export :symbol1 :symbol2))` must be rewritten by splitting them like this:
```lisp
(defpackage :pkgname
  (:use ...))

(defpackage :pkgname
  (:use ...)
  (:export pkgname::symbol1
           pkgname::symbol2))
```
- `defpackage` Keyword arguments `:import-from` and `:shadowing-import-from` are missing but `:import` and `:shadowing-import` are available


## 12. Numbers

* `short-float`


## 19. Filenames

- `:absolute` and `:relative` directory components
- logical pathnames


## 21. Streams

- `broadcast-stream`
- `broadcast-stream-streams`
- `make-broadcast-stream`
- `read-sequence`
- `write-sequence`


## 22. Printer

- `print-unreadable-object`
- some pretty-printing functionality


## 25. Environment

- `describe` doesn't accept the optional argument

**Comment by masinter** [comment]
ultimately the survey of missing features could turn into a set of issues in the https://github.com/orgs/Interlisp/projects/11

There are Issues already -- #609 "Evaluate Compatibility"
#572 Want ANSI Common Lisp

#785 Survey CLTL2 files

**Comment by Anzus** [comment]
The current PCL in the release is not Gregor's final one but the penultimate one, due to the final one needing at least one function that was CLtL2. The final one was working with Bob's CLtL2 build. I need to learn more about how to build a sysout from source to try and do a build with the CLtL2 sources.On Dec 23, 2023 10:25 PM, Larry Masinter ***@***.***> wrote:
ultimately the survey of missing features could turn into a set of issues in the https://github.com/orgs/Interlisp/projects/11
There are Issues already -- #609 "Evaluate Compatibility"
#572 Want ANSI Common Lisp
#785 Survey CLTL2 files

**Comment by pamoroso** [comment]
@masinter Do you mean creating a separate issue in the `common lisp` project for every item in the [list](https://github.com/orgs/Interlisp/discussions/5#discussioncomment-7935648)?

By the way, I added to the list an entry for `describe` which doesn't accept the optional argument.

**Comment by masinter** [comment]
@pamoroso short answer: No, leave them in one file for now. 

Longer anser: 
I don't think we have the resources to complete the task of bringing Medley to ANSI Common Lisp standard.... 
but I really don't know. I think the chart of things-not-implemented yet would be essential in evaluating feasibility.  #609 seems like it would the smallest set of differences to overcome.

**Reply by pamoroso** [reply]
My focus right now is more on documenting what's different in Medley so users like me who want to write Common Lisp code know what isn't available or doesn't work.

**Comment by pamoroso** [comment]
The `#p` reader macro and the `:absolute` and `:relative` directory components are apparently not supported as [reported](https://github.com/Interlisp/medley/issues/609#issuecomment-2003844971) in issue #609.
